File: server/src/api/camera_telemetry.rs
use std::sync::Arc;

use axum::extract::State;
use axum::response::{IntoResponse, Response};
use axum::Extension;
use axum::Json;
use ghostcam::api_types::{CameraCommand, TelemetryPollRequest, TelemetryPollResponse};

use super::auth::AuthCamera;
use super::state::AppState;
use crate::redis::telemetry::write_telemetry_batch;

/// POST /api/v1/cameras/:id/telemetry
///
/// Camera-auth endpoint. Camera posts telemetry every 10s.
/// Response contains any pending commands from the server.
pub async fn post_telemetry(
    State(state): State<Arc<AppState>>,
    Extension(camera): Extension<AuthCamera>,
    Json(body): Json<TelemetryPollRequest>,
) -> Response {
    let device_id = &camera.device_id;

    // Write telemetry to Redis
    if let Some(ref redis) = state.redis {
        write_telemetry_batch(redis, device_id, &[body.telemetry]).await;
    }

    // Claim pending commands
    let commands = match state.db.claim_commands(device_id).await {
        Ok(cmd_values) => cmd_values
            .into_iter()
            .filter_map(|v| serde_json::from_value::<CameraCommand>(v).ok())
            .collect(),
        Err(e) => {
            tracing::warn!(device_id = %device_id, "failed to claim commands: {e}");
            Vec::new()
        }
    };

    Json(TelemetryPollResponse { commands }).into_response()
}

File: server/handlers/camera_telemetry.go
package handlers

import (
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/cargocam/ghostcam/common"
	"github.com/cargocam/ghostcam/server/ctxutil"
	"github.com/cargocam/ghostcam/server/redis"
)

// PostTelemetry handles POST /api/v1/cameras/{deviceID}/telemetry.
func (h *Handlers) PostTelemetry(w http.ResponseWriter, r *http.Request) {
	deviceID := ctxutil.GetCameraDeviceID(r)

	var body common.TelemetryPollRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	// Write telemetry to Redis
	if h.Redis != nil {
		redis.WriteTelemetry(r.Context(), h.Redis.RDB(), deviceID, &body.Telemetry)
	}

	// Mark camera as seen (non-fatal)
	if err := h.DB.TouchCameraLastSeen(r.Context(), deviceID); err != nil {
		slog.Warn("failed to touch camera last_seen_at", "device_id", deviceID, "error", err)
	}

	// Claim pending commands
	commands, err := h.DB.ClaimCommands(r.Context(), deviceID)
	if err != nil {
		// Non-fatal: return empty commands
		commands = nil
	}

	var apiCommands []common.CameraCommand
	for _, raw := range commands {
		var cmd common.CameraCommand
		if json.Unmarshal(raw, &cmd) == nil {
			apiCommands = append(apiCommands, cmd)
		}
	}

	writeJSON(w, http.StatusOK, common.TelemetryPollResponse{Commands: apiCommands})
}
